File: prototype/src/api/contacts.ts
/**
 * ABOUTME: Emergency Contact API module for managing a home's emergency contact list.
 */

import { apiClient } from './client';
import { BackendEmergencyContact, BackendContactListResponse } from './types';
import { EmergencyContact } from '../types';

export const mapBackendContactToContact = (bc: BackendEmergencyContact): EmergencyContact => ({
  id: bc.id,
  homeId: bc.home_id,
  name: bc.name,
  phone: bc.phone,
  relationship: bc.relationship,
  createdAt: bc.created_at,
});

export const listContacts = async (homeId: string): Promise<EmergencyContact[]> => {
  const response = await apiClient.get<BackendContactListResponse>(`/homes/${homeId}/contacts`);
  return response.data.contacts.map(mapBackendContactToContact);
};

export const createContact = async (
  homeId: string, 
  data: Omit<EmergencyContact, 'id' | 'homeId' | 'createdAt'>
): Promise<EmergencyContact> => {
  const response = await apiClient.post<BackendEmergencyContact>(`/homes/${homeId}/contacts`, data);
  return mapBackendContactToContact(response.data);
};

export const patchContact = async (
  homeId: string, 
  contactId: string, 
  data: Partial<Omit<EmergencyContact, 'id' | 'homeId' | 'createdAt'>>
): Promise<EmergencyContact> => {
  const response = await apiClient.patch<BackendEmergencyContact>(`/homes/${homeId}/contacts/${contactId}`, data);
  return mapBackendContactToContact(response.data);
};

export const deleteContact = async (homeId: string, contactId: string): Promise<void> => {
  await apiClient.delete(`/homes/${homeId}/contacts/${contactId}`);
};

File: backend/app/routers/contacts.py
from fastapi import APIRouter, HTTPException, Depends, status
from typing import List
from uuid import UUID
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.schemas.contact import ContactCreate, ContactResponse, ContactListResponse, ContactUpdate
from app.database import get_db
from app.dependencies import get_home_membership
from app.models.emergency_contact import EmergencyContact
from app.models.home_member import HomeMember

router = APIRouter()

@router.post("/{home_id}/contacts", response_model=ContactResponse)
async def create_contact(
    home_id: UUID, 
    data: ContactCreate, 
    member: HomeMember = Depends(get_home_membership),
    db: AsyncSession = Depends(get_db)
):
    new_contact = EmergencyContact(
        home_id=home_id,
        name=data.name,
        phone=data.phone,
        relationship=data.relationship
    )
    db.add(new_contact)
    await db.commit()
    await db.refresh(new_contact)
    return new_contact

@router.get("/{home_id}/contacts", response_model=ContactListResponse)
async def list_contacts(
    home_id: UUID, 
    member: HomeMember = Depends(get_home_membership),
    db: AsyncSession = Depends(get_db)
):
    res = await db.execute(select(EmergencyContact).where(EmergencyContact.home_id == home_id))
    contacts = list(res.scalars().all())
    return ContactListResponse(contacts=contacts)

@router.patch("/{home_id}/contacts/{contact_id}", response_model=ContactResponse)
async def update_contact(
    home_id: UUID, 
    contact_id: UUID, 
    data: ContactUpdate, 
    member: HomeMember = Depends(get_home_membership),
    db: AsyncSession = Depends(get_db)
):
    res = await db.execute(select(EmergencyContact).where(EmergencyContact.id == contact_id, EmergencyContact.home_id == home_id))
    contact = res.scalars().first()
    
    if not contact:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Contact not found")
        
    if data.name is not None:
        contact.name = data.name
    if data.phone is not None:
        contact.phone = data.phone
    if data.relationship is not None:
        contact.relationship = data.relationship
        
    await db.commit()
    await db.refresh(contact)
    return contact

@router.delete("/{home_id}/contacts/{contact_id}", response_model=dict)
async def delete_contact(
    home_id: UUID, 
    contact_id: UUID, 
    member: HomeMember = Depends(get_home_membership),
    db: AsyncSession = Depends(get_db)
):
    res = await db.execute(select(EmergencyContact).where(EmergencyContact.id == contact_id, EmergencyContact.home_id == home_id))
    contact = res.scalars().first()
    if not contact:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Contact not found")
        
    await db.delete(contact)
    await db.commit()
    return {"message": "Contact deleted successfully"}
